File: solutions/elixir/house/3/lib/house.ex
defmodule House do
  @doc """
  Return verses of the nursery rhyme 'This is the House that Jack Built'.
  """
  @verses """
          the house that Jack built.
          the malt that lay in
          the rat that ate
          the cat that killed
          the dog that worried
          the cow with the crumpled horn that tossed
          the maiden all forlorn that milked
          the man all tattered and torn that kissed
          the priest all shaven and shorn that married
          the rooster that crowed in the morn that woke
          the farmer sowing his corn that kept
          the horse and the hound and the horn that belonged to
          """
          |> String.split("\n")

  @spec recite(start :: integer, stop :: integer) :: String.t()
  def recite(start, stop) do
    Enum.map_join(start..stop, "", fn i ->
      "This is #{do_verses(i)}"
    end)
  end

  defp do_verses(top) do
    @verses
    |> Enum.take(top)
    |> Enum.reverse()
    |> Enum.join(" ")
    |> Kernel.<>("\n")
  end
end

File: solutions/typescript/house/1/house.ts
const VERSES: string[] = `the house that Jack built.
the malt\tthat lay in
the rat\tthat ate
the cat\tthat killed
the dog\tthat worried
the cow with the crumpled horn\tthat tossed
the maiden all forlorn\tthat milked
the man all tattered and torn\tthat kissed
the priest all shaven and shorn\tthat married
the rooster that crowed in the morn\tthat woke
the farmer sowing his corn\tthat kept
the horse and the hound and the horn\tthat belonged to`.split("\n");


export function verse(maxVerse: number): string[] {
  let ret: string[] = VERSES.slice(0, maxVerse)
    .reverse()
    .join(" ")
    .replaceAll("\t", "\n").split("\n");
  ret[0] = `This is ${ret[0]}`;
  return ret;
}

export function verses(startNum: number, stopNum: number): string[] {
  return [...Array(stopNum - startNum + 1).keys()]
    .flatMap((num: number): string[] => [...verse(num + startNum), ''])
    .slice(0, -1);
}
